packages/core/src/common/messaging/abstract-connection-provider.ts

```diff
@@ -0,0 +1,122 @@
+/********************************************************************************
+ * Copyright (C) 2020 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, interfaces } from 'inversify';
+import { ConsoleLogger, createWebSocketConnection, Logger } from 'vscode-ws-jsonrpc';
+import { Emitter, Event } from '../event';
+import { ConnectionHandler } from './handler';
+import { JsonRpcProxy, JsonRpcProxyFactory } from './proxy-factory';
+import { WebSocketChannel } from './web-socket-channel';
+
+/**
+ * Factor common logic according to `ElectronIpcConnectionProvider` and
+ * `WebSocketConnectionProvider`. This class handles channels in a somewhat
+ * generic way.
+ */
+@injectable()
+export abstract class AbstractConnectionProvider<AbstractOptions extends object> {
+
+    /**
+     * Create a proxy object to remote interface of T type
+     * over an electron ipc connection for the given path and proxy factory.
+     */
+    static createProxy<T extends object>(container: interfaces.Container, path: string, factory: JsonRpcProxyFactory<T>): JsonRpcProxy<T>;
+    /**
+     * Create a proxy object to remote interface of T type
+     * over an electron ipc connection for the given path.
+     *
+     * An optional target can be provided to handle
+     * notifications and requests from a remote side.
+     */
+    static createProxy<T extends object>(container: interfaces.Container, path: string, target?: object): JsonRpcProxy<T> {
```

Comment:
I don't think we need code which does nothing at runtime. It should be rather removed and a proper method added to each provider.

Reply:
The class is abstract, and this is a workaround to having abstract static methods... Subclasses provide the actual implementation for what are helper functions anyway (doing `SomeConnectionProvider.createProxy(...)`).

This simplifies Inversify bindings, and respects the old API to not break clients.

Reply:
Else we need to copy/paste the signature overloads + doc in each subclass.

Reply:
I am fine with both. I leave as is.